src/robot/variables/tablesetter.py

```diff
@@ -75,15 +88,26 @@ def resolve(self, variables) -> Any:
         if not self.resolved:
             self.resolving = True
             try:
-                self.value = self._replace_variables(variables)
+                value = self._replace_variables(variables)
             finally:
                 self.resolving = False
+            self.value = self._convert(value, self.type) if self.type else value
+            if self.name:
+                self.name = self.name[:2] + variables.replace_string(self.name[2:-1]) + '}'
             self.resolved = True
         return self.value
 
     def _replace_variables(self, variables) -> Any:
         raise NotImplementedError
 
+    def _convert(self, value, type_):
+        from robot.running import TypeInfo
+        info = TypeInfo.from_type_hint(type_)
```

Comment:
Using `TypeInfo.from_variable` would avoid the need to handle `@` and `&` variables in `ListVariableResolver` and `DictVariableResolver`, but because we already have a type here, and needed to construct a variable, that might be more complicated anyway.